### PR TITLE
Refactoring phase for tessellation factor in TCS

### PR DESCRIPTION
**Pull request by xuechen417**

### CODE REVIEW

**Comment by amdvlk-admin** [comment]
### Test summary for commit ad3660feb5a3e5400ef958179229c7a83ff603d9
<details><summary>Driver commits used in build</summary><ul>

<li>CWPACK: amd-master 7387247eb9889ddcabbc1053b9c2052e253b088e</li>
<li>METROHASH: amd-master 3c566dd9cda44ca7fd97659e0b53ac953f9037d2</li>
<li>PAL: dev eb28c7ef53329f838e693f233696823330e964c0</li>
<li>SPVGEN: dev e1f1881d84cd1c7a6c1d661cbac0e619b4e110c1</li>
<li>XGL: dev 63761ecfd2d3cfa61a756ce0a72e1ca1cdae333a</li>
<li>LLVM-PROJECT: amd-gfx-gpuopen-dev ae9a45d0cf074c06dad804b42927178d7c5f3849</li>
</ul></details>

<details><summary>CTS tests (Failed: 483/144501)</summary><ul><li>Built with version 1.2.2.2</li>
<details open><summary>Rhel 8.2, Gfx10</summary><ul>

<li>Passed: 25846/48174 (53.7%)</li>
<li><details><summary>Failed: 481/48174 (1.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.api.buffer.suballocation.storage.vertex.create.binding
Stack trace: Flake

FAILURE: dEQP-VK.api.buffer.suballocation.transfer_dst.uniform_texel.storage_texel.uniform.storage.vertex.create.binding_residency
Stack trace: Flake

FAILURE: dEQP-VK.api.buffer.suballocation.transfer_dst.uniform_texel.uniform.storage.vertex.indirect.create.binding_residency_aliased
Stack trace: Flake

FAILURE: dEQP-VK.api.buffer.suballocation.transfer_src.transfer_dst.uniform_texel.storage_texel.uniform.index.create.binding
Stack trace: Flake

FAILURE: dEQP-VK.api.copy_and_blit.core.blit_image.all_formats.color.r32_sfloat.a8b8g8r8_unorm_pack32.general_optimal_nearest
Stack trace: Flake

FAILURE: dEQP-VK.api.copy_and_blit.core.image_to_image.all_formats.color.r32_sfloat.a8b8g8r8_uint_pack32.optimal_general
Stack trace: Flake

FAILURE: dEQP-VK.api.copy_and_blit.dedicated_allocation.blit_image.all_formats.color.r4g4_unorm_pack8.r5g6b5_unorm_pack16.general_linear_nearest
Stack trace: Flake

FAILURE: dEQP-VK.api.copy_and_blit.dedicated_allocation.blit_image.all_formats.generate_mipmaps.from_previous_level.layercount_1.r8g8b8a8_srgb.general_optimal_linear
Stack trace: Flake

FAILURE: dEQP-VK.api.copy_and_blit.dedicated_allocation.image_to_image.dimensions.src4096x4_dst4096x4.r4g4_unorm_pack8.r8_srgb.optimal_general
Stack trace: Flake

FAILURE: dEQP-VK.api.fill_and_update_buffer.suballocation.fill_buffer_first_one
Stack trace: Flake

FAILURE: dEQP-VK.api.image_clearing.core.clear_color_attachment.single_layer.r5g5b5a1_unorm_pack16_33x128
Stack trace: Flake

FAILURE: dEQP-VK.api.image_clearing.core.clear_color_image.1d.linear.single_layer.e5b9g9r9_ufloat_pack32_71x1
Stack trace: Flake

FAILURE: dEQP-VK.api.image_clearing.core.clear_color_image.2d.optimal.multiple_layers.a2b10g10r10_uint_pack32_200x180
Stack trace: Flake

FAILURE: dEQP-VK.api.image_clearing.core.clear_color_image.2d.optimal.remaining_array_layers.r8_sint_64x11_multiple_subresourcerange
Stack trace: Flake

FAILURE: dEQP-VK.api.image_clearing.core.clear_color_image.3d.optimal.single_layer.r16g16b16a16_sfloat
Stack trace: Flake

FAILURE: dEQP-VK.api.image_clearing.dedicated_allocation.clear_color_attachment.single_layer.r8_sint_64x11
Stack trace: Flake

FAILURE: dEQP-VK.api.image_clearing.dedicated_allocation.clear_color_image.2d.linear.multiple_layers.r4g4_unorm_pack8_1x33
Stack trace: Flake

FAILURE: dEQP-VK.api.image_clearing.dedicated_allocation.clear_color_image.2d.linear.remaining_array_layers.a2r10g10b10_unorm_pack32
Stack trace: Flake

FAILURE: dEQP-VK.api.image_clearing.dedicated_allocation.clear_color_image.2d.linear.remaining_array_layers.r32g32b32a32_sfloat_33x128
Stack trace: Flake

FAILURE: dEQP-VK.api.image_clearing.dedicated_allocation.clear_color_image.2d.linear.remaining_array_layers_twostep.b5g6r5_unorm_pack16_64x11
Stack trace: Flake
...
```
</p></details></li>

<li>Skipped: 21847/48174 (45.4%)</li>
</ul></details>

<details open><summary>Ubuntu 18.04, Gfx9</summary><ul>

<li>Passed: 26319/48174 (54.6%)</li>
<li><details><summary>Failed: 1/48174 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.graphicsfuzz.cov-fold-div-variable-by-zero
Stack trace: Fail
```
</p></details></li>

<li>Skipped: 21854/48174 (45.4%)</li>
</ul></details>

<details open><summary>Ubuntu 20.04, Gfx8</summary><ul>

<li>Passed: 24851/48153 (51.6%)</li>
<li><details><summary>Failed: 1/48153 (0.0%)</summary><p>


Failures:
```
FAILURE: dEQP-VK.graphicsfuzz.cov-fold-div-variable-by-zero
Stack trace: Fail
```
</p></details></li>

<li>Skipped: 23301/48153 (48.4%)</li>
</ul></details>


</ul></details>

File: lgc/patch/PatchInOutImportExport.cpp [hunk 1]
```diff
@@ -2783,120 +2840,13 @@ void PatchInOutImportExport::patchTcsBuiltInOutputExport(Value *output, unsigned
...
-          auto elemIdx = ConstantInt::get(Type::getInt32Ty(*m_context), i);
-          auto ldsOffset = calcLdsOffsetForTcsOutput(elem->getType(), loc, nullptr, elemIdx, vertexIdx, insertPos);
-          writeValueToLds(elem, ldsOffset, insertPos);
-        }
-      } else {
-        auto ldsOffset = calcLdsOffsetForTcsOutput(outputTy, loc, nullptr, elemIdx, nullptr, insertPos);
-        writeValueToLds(output, ldsOffset, insertPos);
-      }
-    }
+    if (builtInUsage.tessLevelOuter)
```
Comment (nhaehnle) [inline lgc/patch/PatchInOutImportExport.cpp:2843]: This isn't specific to your change, but why do we have this check here?

The function is responsible for writing TCS output builtins, and `builtInUsage` is the usage of TCS. So if this function is called for `builtInId == BuiltInTessLevelOuter`, then shouldn't `builtInUsage.tessLevelOuter` **always** be true here?

The same logic applies to BuiltInTessLevelInner.

If my reasoning here is correct, then you can remove the if-check and make the code slightly easier to follow by moving the lambda function back here into the switch, like:
```
case BuiltInTessLevelOuter:
case BuiltInTessLevelInner:{
  // Extract tessellation factors
  bool isOuter = builtInId == BuiltInTessLevelOuter;
  std::vector<Value *> tessFactors;
  if (!elemIdx) {
    assert(outputTy->isArrayTy());
    unsigned tessFactorCount = 0;
    const unsigned adjust = isOuter ? 2 : 0;
    ...
```
This allows us to read the code without jumping around so much.

**Review by nhaehnle — COMMENTED** [review]
Thank you for splitting the two patches and the good commit messages, those really help!

I do have one comment on the first patch which could perhaps improve it even more.

The second patch already looks great to me!

File: lgc/patch/PatchInOutImportExport.cpp (hunk 1 above)
Comment (nhaehnle) [reply lgc/patch/PatchInOutImportExport.cpp:2843]: I see that you do a related change in #1089. I think removing the lambda function and doing everything in the switch statement would really be better.

Comment (xuechen417) [reply lgc/patch/PatchInOutImportExport.cpp:2843]: Yes, you're right. The check for `builtInUsage.tessLevelOuter` is unnecessary. Removed the checks and lambda function.

**Review by xuechen417 — COMMENTED** [review]
V1: Removed unnecessary checks and the lambda function for extracted common code.

**Comment by xuechen417** [comment]
PR #1089  is based on this PR including all generic changes for TFs. Considering the new review comments are made in PR #1089, I closed this PR and use PR #1089 for review and merge.

**Comment by kuhar** [comment]
@xuechen417  Sorry, I noticed that these PRs have each multiple commits, but didn't realize this is the base one. I can never remember if github shows commits in chronological or reverse-chronological order and got it confused...

**Comment by xuechen417** [comment]
> @xuechen417 Sorry, I noticed that these PRs have each multiple commits, but didn't realize this is the base one. I can never remember if github shows commits in chronological or reverse-chronological order and got it confused...

Never mind. Keep one PR #1089 to track is more convenient for review.

**Pull request closed without merging**